src/notebooks/controllers/controllerRegistration.ts

```diff
@@ -0,0 +1,190 @@
+
+    public get onCreated(): Event<IVSCodeNotebookController> {
+        return this.creationEmitter.event;
+    }
+    public get values(): IVSCodeNotebookController[] {
+        return [...this.registeredControllers.values()];
+    }
+    public get connections(): KernelConnectionMetadata[] {
+        return [...this.registeredConnections.values()];
+    }
+    constructor(
+        @inject(IVSCodeNotebook) private readonly notebook: IVSCodeNotebook,
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        @inject(IConfigurationService) private readonly configuration: IConfigurationService,
+        @inject(ICommandManager) private readonly commandManager: ICommandManager,
+        @inject(KernelFilterService) private readonly kernelFilter: KernelFilterService,
+        @inject(IExtensionContext) private readonly context: IExtensionContext,
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider,
+        @inject(NotebookCellLanguageService) private readonly languageService: NotebookCellLanguageService,
+        @inject(IWorkspaceService) private readonly workspace: IWorkspaceService,
+        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker,
+        @inject(IDocumentManager) private readonly docManager: IDocumentManager,
+        @inject(IApplicationShell) private readonly appShell: IApplicationShell,
+        @inject(IBrowserService) private readonly browser: IBrowserService,
+        @inject(IServiceContainer) private readonly serviceContainer: IServiceContainer,
+        @inject(ILocalResourceUriConverter) private readonly resourceConverter: ILocalResourceUriConverter
+    ) {
+        this.kernelFilter.onDidChange(this.onDidChangeKernelFilter, this, this.disposables);
+    }
+    add(
+        metadata: KernelConnectionMetadata,
+        types: ('jupyter-notebook' | 'interactive')[]
+    ): IVSCodeNotebookController[] {
+        let results: IVSCodeNotebookController[] = [];
+        try {
+            // Create notebook selector
+            types
+                .map((t) => {
+                    return [this.getControllerId(metadata, t), t];
+                })
+                .filter(([id]) => {
+                    // Update our list kernel connections.
+                    this.registeredConnections.set(id, metadata);
+
+                    // See if we already created this controller or not
+                    const controller = this.registeredControllers.get(id);
+                    if (controller) {
+                        // If we already have this controller, its possible the Python version information has changed.
+                        // E.g. we had a cached kernlespec, and since then the user updated their version of python,
+                        // Now we need to update the display name of the kernelspec.
+                        controller.updateConnection(metadata);
```

Comment:
One of the changes made was to always update existing controllers if the metadata id is a match. We were only doing this for specific cases before but I figured why not just always update it.

Reply:
Yeah, i guess it makes it siimpler to read and maintain